File: sections/76_suffix_arrays_lcp/problems/d_disjoint_repeated_substring/solution.cpp
#include <bits/stdc++.h>
using namespace std;
vector<int> suffix_array(const string &s) {
  int n = s.size();
  vector<int> sa(n), r(n), nr(n);
  iota(sa.begin(), sa.end(), 0);
  for (int i = 0; i < n; ++i)
    r[i] = (unsigned char)s[i];
  for (int k = 1;; k *= 2) {
    sort(sa.begin(), sa.end(), [&](int a, int b) {
      return pair{r[a], a + k < n ? r[a + k] : -1} <
             pair{r[b], b + k < n ? r[b + k] : -1};
    });
    nr[sa[0]] = 0;
    for (int i = 1; i < n; ++i) {
      int a = sa[i - 1], b = sa[i];
      nr[b] = nr[a] + (pair{r[a], a + k < n ? r[a + k] : -1} <
                       pair{r[b], b + k < n ? r[b + k] : -1});
    }
    r = nr;
    if (r[sa.back()] == n - 1)
      break;
  }
  return sa;
}
vector<int> lcp_array(const string &s, const vector<int> &sa) {
  int n = s.size(), h = 0;
  vector<int> rank(n), lcp(max(0, n - 1));
  for (int i = 0; i < n; ++i)
    rank[sa[i]] = i;
  for (int i = 0; i < n; ++i) {
    int q = rank[i];
    if (!q)
      continue;
    int j = sa[q - 1];
    while (i + h < n && j + h < n && s[i + h] == s[j + h])
      ++h;
    lcp[q - 1] = h;
    if (h)
      --h;
  }
  return lcp;
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  string s;
  cin >> s;
  int n = s.size();
  auto sa = suffix_array(s), lcp = lcp_array(s, sa);
  auto ok = [&](int need) {
    int lo = sa[0], hi = sa[0];
    for (int i = 0; i < n - 1; ++i) {
      if (lcp[i] < need)
        lo = hi = sa[i + 1];
      else {
        lo = min(lo, sa[i + 1]);
        hi = max(hi, sa[i + 1]);
        if (hi - lo >= need)
          return true;
      }
    }
    return false;
  };
  int l = 0, r = n / 2 + 1;
  while (r - l > 1) {
    int m = (l + r) / 2;
    (ok(m) ? l : r) = m;
  }
  cout << l << '\n';
}

File: sections/76_suffix_arrays_lcp/problems/d_disjoint_repeated_substring/solution.py
def suffix_array(s):
    n = len(s)
    sa = list(range(n))
    rank = list(map(ord, s))
    step = 1
    while True:
        sa.sort(
            key=lambda i: (rank[i], rank[i + step] if i + step < n else -1)
        )
        new_rank = [0] * n
        for j in range(1, n):
            a, b = sa[j - 1], sa[j]
            left = (rank[a], rank[a + step] if a + step < n else -1)
            right = (rank[b], rank[b + step] if b + step < n else -1)
            new_rank[b] = new_rank[a] + (left < right)
        rank = new_rank
        if rank[sa[-1]] == n - 1:
            return sa
        step *= 2


def lcp_array(s, sa):
    n = len(s)
    rank = [0] * n
    for i, start in enumerate(sa):
        rank[start] = i
    lcp = [0] * (n - 1)
    height = 0
    for i in range(n):
        place = rank[i]
        if place == 0:
            continue
        j = sa[place - 1]
        while (
            i + height < n
            and j + height < n
            and s[i + height] == s[j + height]
        ):
            height += 1
        lcp[place - 1] = height
        height = max(0, height - 1)
    return lcp


def main():
    import sys

    s = sys.stdin.buffer.readline().decode().strip()
    sa = suffix_array(s)
    lcp = lcp_array(s, sa)

    def possible(length):
        low = high = sa[0]
        for i, common in enumerate(lcp):
            if common < length:
                low = high = sa[i + 1]
            else:
                low = min(low, sa[i + 1])
                high = max(high, sa[i + 1])
                if high - low >= length:
                    return True
        return False

    low, high = 0, len(s) // 2 + 1
    while high - low > 1:
        middle = (low + high) // 2
        if possible(middle):
            low = middle
        else:
            high = middle
    print(low)


if __name__ == "__main__":
    main()
